modules/ingest-common/src/main/java/org/opensearch/ingest/common/RemoveByPatternProcessor.java

```diff
@@ -0,0 +1,180 @@
+            throw new IllegalArgumentException("either fieldPatterns and excludeFieldPatterns must be set");
+        }
+        if (fieldPatterns == null) {
+            this.fieldPatterns = null;
+            this.excludeFieldPatterns = new ArrayList<>(excludeFieldPatterns);
+        } else {
+            this.fieldPatterns = new ArrayList<>(fieldPatterns);
+            this.excludeFieldPatterns = null;
+        }
+    }
+
+    public List<String> getFieldPatterns() {
+        return fieldPatterns;
+    }
+
+    public List<String> getExcludeFieldPatterns() {
+        return excludeFieldPatterns;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument document) {
+        Set<String> existingFields = new HashSet<>(document.getSourceAndMetadata().keySet());
+        Set<String> metadataFields = document.getMetadata()
+            .keySet()
+            .stream()
+            .map(IngestDocument.Metadata::getFieldName)
+            .collect(Collectors.toSet());
+
+        if (fieldPatterns != null && !fieldPatterns.isEmpty()) {
+            existingFields.forEach(field -> {
+                // ignore metadata fields such as _index, _id, etc.
+                if (!metadataFields.contains(field)) {
+                    final boolean matched = fieldPatterns.stream().anyMatch(pattern -> Regex.simpleMatch(pattern, field));
+                    if (matched) {
+                        document.removeField(field);
+                    }
+                }
+            });
+        }
+
+        if (excludeFieldPatterns != null && !excludeFieldPatterns.isEmpty()) {
+            existingFields.forEach(field -> {
+                // ignore metadata fields such as _index, _id, etc.
+                if (!metadataFields.contains(field)) {
+                    final boolean matched = excludeFieldPatterns.stream().anyMatch(pattern -> Regex.simpleMatch(pattern, field));
+                    if (!matched) {
+                        document.removeField(field);
+                    }
+                }
+            });
+        }
+
+        return document;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    public static final class Factory implements Processor.Factory {
+
+        public Factory() {}
+
+        @Override
+        public RemoveByPatternProcessor create(
+            Map<String, Processor.Factory> registry,
+            String processorTag,
+            String description,
+            Map<String, Object> config
+        ) throws Exception {
+            final List<String> fieldPatterns = new ArrayList<>();
+            final List<String> excludeFieldPatterns = new ArrayList<>();
+            final Object fieldPattern = ConfigurationUtils.readOptionalObject(config, "field_pattern");
+            final Object excludeFieldPattern = ConfigurationUtils.readOptionalObject(config, "exclude_field_pattern");
```

Comment:
Overall, the change looks good.My only nitpick is the name of this parameter. 

I get the idea that it's a `remove` processor and we're excluding fields from being removed, but I can also see how someone might see a processor configured with just `exclude_field_pattern` and think it's talking about the fields that are "excluded" from the final document.

What do you think of `keep_field_pattern` or `keep_only_field_pattern`? I feel like those might better convey that while it it's a `remove_by_pattern` processor, fields that match *this* pattern will be kept. Naming is always hard.

Reply:
Remove processor now has a `exclude_field` parameter which is used to retain fields in the list and remove others, I think we should make the naming consistent with that. @reta, what do you think about this?

Reply:
Oh -- I didn't realize that `remove` already had that!

In that case, I agree that consistency is more important. I don't particularly like the name, but we should definitely be consistent. :+1: